File: application/app/domain/services/EndpointsProvider.scala
package domain.services

import domain.exceptions.EndpointNotFoundException
import domain.model.Endpoint
import domain.repositories.EndpointRepository
import javax.inject.{Inject, Named, Singleton}

import scala.concurrent._
import ExecutionContext.Implicits.global

@Singleton
class EndpointsProvider @Inject() (@Named("endpointRepo") endpointRepository: EndpointRepository) {
  def getFirstMatch(path: String): Future[Option[Endpoint]] = getAll()
    .map(endpoints => endpoints.find(e => e.matchesPath(path)))

  def getAll(): Future[Seq[Endpoint]] = endpointRepository.getAll()

  def get(id: String): Future[Endpoint] = endpointRepository.getById(id)
    .map(getEndpointOrThrow)

  private def getEndpointOrThrow(maybeEndpoint: Option[Endpoint]): Endpoint = maybeEndpoint match {
    case Some(e) => e
    case None => throw EndpointNotFoundException(s"Endpoint not found")
  }
}

File: ws-gateway-application/src/main/java/com/cosmin/wsgateway/application/services/EndpointsProvider.java
package com.cosmin.wsgateway.application.services;

import com.cosmin.wsgateway.application.repositories.EndpointRepository;
import com.cosmin.wsgateway.domain.exceptions.EndpointNotFoundException;
import com.cosmin.wsgateway.domain.model.Endpoint;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;

@Component
@RequiredArgsConstructor
public class EndpointsProvider {
    private final EndpointRepository endpointRepository;

    public Flux<Endpoint> getAll() {
        return endpointRepository.getAll();
    }

    public Mono<Endpoint> getById(String id) {
        return endpointRepository.getById(id)
                .map(e -> e.orElseThrow(() -> new EndpointNotFoundException(id)));
    }

    public Mono<Endpoint> getFirstMatch(String path) {
        return endpointRepository.getAll()
                .collectList()
                .map(endpoints -> getEndpointThatMatchesPath(path, endpoints));
    }

    private Endpoint getEndpointThatMatchesPath(String path, List<Endpoint> endpoints) {
        return endpoints.stream()
                .filter(e -> e.matchesPath(path))
                .findFirst()
                .orElseThrow(() -> new EndpointNotFoundException("path"));
    }
}
